### PR TITLE
[BUG][Vectorized] fix `replace_if_not_null` in vectorized compaction

### LINKED ISSUE #9374: [Bug] `replace_if_not_null` not found in vectorized compaction
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Version

master branch

### What's Wrong?

Some thing wrong for `REPLACE NOT NULL` aggregation type column in vectorized compaction now.

`
 1# 0x00007FE7DDEC6400 in /lib64/libc.so.6
 2# doris::vectorized::BlockReader::_init_agg_state(doris::TabletReader::ReaderParams const&) at /root/incubator-doris/be/src/vec/olap/block_reader.cpp:107
 3# doris::vectorized::BlockReader::init(doris::TabletReader::ReaderParams const&) at /root/incubator-doris/be/src/vec/olap/block_reader.cpp:162
 4# doris::Merger::vmerge_rowsets(std::shared_ptr<doris::Tablet>, doris::ReaderType, std::vector<std::shared_ptr<doris::RowsetReader>, std::allocator<std::shared_ptr<doris::RowsetReader> > > const&, doris::RowsetWriter*, doris::Merger::Statistics*) at /root/incubator-doris/be/src/olap/merger.cpp:106
 5# doris::Compaction::do_compaction_impl(long) at /root/incubator-doris/be/src/olap/compaction.cpp:95
 6# doris::Compaction::do_compaction(long) at /root/incubator-doris/be/src/olap/compaction.cpp:57
 7# doris::CumulativeCompaction::execute_compact_impl() at /root/incubator-doris/be/src/olap/cumulative_compaction.cpp:75
 8# doris::Compaction::execute_compact() at /root/incubator-doris/be/src/olap/compaction.cpp:46
 9# doris::Tablet::execute_compaction(doris::CompactionType) at /root/incubator-doris/be/src/olap/tablet.cpp:1407
10# std::_Function_handler<void (), doris::StorageEngine::_submit_compaction_task(std::shared_ptr<doris::Tablet>, doris::CompactionType)::{lambda()#1}>::_M_invoke(std::_Any_data const&) at /var/local/ldb-toolchain/include/c++/11/bits/std_function.h:291
11# doris::ThreadPool::dispatch_thread() at /root/incubator-doris/be/src/util/threadpool.cpp:548
12# doris::Thread::supervise_thread(void*) at /root/incubator-doris/be/src/util/thread.cpp:409
13# start_thread in /lib64/libpthread.so.0
14# clone in /lib64/libc.so.6

`

### What You Expected?

fix it

### How to Reproduce?

_No response_

### Anything Else?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Gabriel39**
# Proposed changes

Issue Number: close #9374 

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/aggregate_functions/aggregate_function_window.h [hunk 1]
```diff
@@ -348,6 +353,44 @@ struct WindowFunctionFirstData : Data {
...
+        }
+        if (frame_start < frame_end &&
+            frame_end <= partition_start) { //rewrite last_value when under partition
+            this->set_is_null();            //so no need more judge
+            return;
+        }
+        frame_start = std::max<int64_t>(frame_start, partition_start);
+        frame_end = std::min<int64_t>(frame_end, partition_end);
+        for (int i = frame_start; i < frame_end; i++) {
+            if (const auto* nullable_column = check_and_get_column<ColumnNullable>(columns[0])) {
```
Comment (BiteTheDDDDt) [inline be/src/vec/aggregate_functions/aggregate_function_window.h:371]: Why not move check_and_get_column outside the loop to improve performance？

Comment (Gabriel39) [reply be/src/vec/aggregate_functions/aggregate_function_window.h]: Done, thanks for remind

File: be/src/vec/aggregate_functions/aggregate_function_window.h [hunk 2]
```diff
@@ -192,6 +194,9 @@ struct LeadAndLagData {
         _default_value.reset();
         _is_init = false;
         _has_value = false;
+        if constexpr (init_with_null) {
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_window.h:197]: the logic only use in child class， should not do in parent.

File: be/src/vec/aggregate_functions/aggregate_function_window.h [hunk 3]
```diff
@@ -408,61 +489,78 @@ class WindowFunctionData final
...
+        return AggregateFunctionPtr(
+                create_function_single_value<WindowFunctionData, WindowFunctionFirstData,
+                                             is_nullable, is_copy>(name, argument_types,
+                                                                   parameters));
+    }
 }
 
-template <bool is_nullable, bool is_copy>
+template <bool is_nullable, bool is_copy, bool replace_if_not_null = false>
 AggregateFunctionPtr create_aggregate_function_last(const std::string& name,
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_window.h:549]: maybe should split the register `last_value` and `last_not_null_value`

Comment (Gabriel39) [reply be/src/vec/aggregate_functions/aggregate_function_window.h:553]: done

File: be/src/vec/aggregate_functions/aggregate_function_window.h [hunk 4]
```diff
@@ -348,6 +355,48 @@ struct WindowFunctionFirstData : Data {
...
+        }
+        if (frame_start < frame_end &&
+            frame_end <= partition_start) { //rewrite last_value when under partition
+            this->set_is_null();            //so no need more judge
+            return;
+        }
+        this->set_null_if_need();
+        frame_start = std::max<int64_t>(frame_start, partition_start);
+        frame_end = std::min<int64_t>(frame_end, partition_end);
+        if (const auto* nullable_column = check_and_get_column<ColumnNullable>(columns[0])) {
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_window.h:373]: use if constexpr to speed up the function

**Review by HappenLee — DISMISSED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**